modules/core/src/main/java/org/apache/ignite/internal/binary/builder/BinaryBuilderReader.java

```diff
@@ -516,15 +516,26 @@ public Object parseValue() {
             case GridBinaryMarshaller.HANDLE: {
                 int objStart = pos - 1 - readInt();
 
-                BinaryObjectBuilderImpl res = objMap.get(objStart);
+                BinaryObjectBuilderImpl bobRef = objMap.get(objStart);
 
-                if (res == null) {
-                    res = new BinaryObjectBuilderImpl(new BinaryBuilderReader(this, objStart), objStart);
+                if (bobRef != null)
+                    return bobRef;
 
-                    objMap.put(objStart, res);
-                }
+                // Read handle by position
+                int savedPos = pos;
+                pos = objStart;
 
-                return res;
+                try {
+                    Object res = parseValue();
+
+                    if (res instanceof BinaryObjectBuilderImpl)
+                        objMap.put(objStart, (BinaryObjectBuilderImpl)res);
+
+                    return res;
+                }
+                finally {
```

Comment:
I'm not sure that we should save 'pos' in Exception case, maybe enough restore 'pos' before return statement?